### PR TITLE
split reader and writer

### PR DESCRIPTION
**Pull request by JanEggers**
+ less generics 
+ more flexibility for protocol negotiation.
(for example in mqtt I need to read the beginning of the connect packet to know what version is used, right now there is a indirection that I hope to remove)

- with this pattern one could try to create multiple writers making the semaphore in the writer useless. 

sample usage:

https://github.com/JanEggers/MQTTnet/blob/5281e5316451b51b739ab229b9e8b2b911f67319/Source/MQTTnet.AspnetCore/Extensions/ConnectionContextExtensions.cs#L11-L18

https://github.com/JanEggers/MQTTnet/blob/5281e5316451b51b739ab229b9e8b2b911f67319/Source/MQTTnet.AspnetCore/MqttConnectionHandler.cs#L29-L32

### CODE REVIEW

File: Bedrock.Framework/Protocols/Protocol.cs [hunk 1]
```diff
@@ -1,27 +1,58 @@
...
 
 namespace Bedrock.Framework.Protocols
 {
-    public class Protocol<TReader, TWriter, TReadMessage, TWriteMessage> where TReader : IProtocolReader<TReadMessage>
-                                                                         where TWriter : IProtocolWriter<TWriteMessage>
+    public static class Protocol 
     {
-        private readonly TReader _reader;
-        private readonly TWriter _writer;
+        public static ProtocolWriter<TWriteMessage> WriteProtocol<TWriteMessage>(this ConnectionContext connection, IProtocolWriter<TWriteMessage> writer)
```
Comment (davidfowl) [inline Bedrock.Framework/Protocols/Protocol.cs]: CreateWriter

File: Bedrock.Framework/Protocols/Protocol.cs [hunk 2]
```diff
@@ -1,27 +1,58 @@
...
+            => new ProtocolWriter<TWriteMessage>(connection, writer);
+
+        public static ProtocolReader<TReadMessage> ReadProtocol<TReadMessage>(this ConnectionContext connection, IProtocolReader<TReadMessage> reader, int? maximumMessageSize = null)
```
Comment (davidfowl) [inline Bedrock.Framework/Protocols/Protocol.cs]: CreateReader

**Comment by davidfowl** [comment]
Do those renames, but this change looks good.

**Comment by JanEggers** [comment]
oki will do,

**Pull request merged**